### DISCUSSION TITLE
Build failed for Ruby 2.7.5 on M1 Mac (Ventura 13.0)

### DISCUSSION THREAD

**Discussion by dungtran3864** [rbenv/ruby-build#2082] [Build failures] [answered]
I'm currently using MacBook Pro 2021 (M1 Pro). My MacOS version is Ventura 13.0. I'm unable to install Ruby 2.7.5 on my machine. Please guide me. Here are the logs for the failed build.

```
To follow progress, use 'tail -f /var/folders/yf/8x2cgk3s0yb58rkxcq5d1p7m0000gn/T/ruby-build.20221107095056.30615.log' or pass --verbose
Downloading ruby-2.7.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.5.tar.bz2
Installing ruby-2.7.5...
ruby-build: using readline from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.0 using ruby-build 20221101)

Inspect or clean up the working tree at /var/folders/yf/8x2cgk3s0yb58rkxcq5d1p7m0000gn/T/ruby-build.20221107095056.30615.EQjeNt
Results logged to /var/folders/yf/8x2cgk3s0yb58rkxcq5d1p7m0000gn/T/ruby-build.20221107095056.30615.log

Last 10 log lines:
make[1]: *** [ext/readline/all] Error 2
make[1]: *** Waiting for unfinished jobs....
compiling date_strptime.c
linking shared-object objspace.bundle
installing default console libraries
installing default date_core libraries
linking shared-object io/console.bundle
linking shared-object nkf.bundle
linking shared-object date_core.bundle
make: *** [build-ext] Error 2
```

Note: I have followed the instructions here for Ruby 2.7.5: https://github.com/rbenv/ruby-build/wiki#suggested-build-environment with all the latest versions of rbenv, ruby-build, readline, gmp, libyml, etc.

I install using command "rbenv install 2.7.5"

Original logs attached in the comment below!

**Comment by mislav** [comment]
Hi, something weird happened to your log file. It looks like some tool edited it or added it after ruby-build initially generated it. Please upload the log file unmodified.

**Comment by dungtran3864** [comment, marked as answer]
@mislav Here are the original logs

[ruby-build.20221107095056.30615.log](https://github.com/rbenv/ruby-build/files/9952804/ruby-build.20221107095056.30615.log)

**Reply by mislav** [reply]
@dungtran3864 Typically Homebrew on M1 is under `/opt/homebrew`. I see your readline being under `/usr/local/opt/readline`. What is your `brew --prefix readline`?

**Reply by dungtran3864** [reply]
@mislav /usr/local/opt/readline

**Reply by dungtran3864** [reply]
I don't see /homebrew in my /usr/local/opt directory. And it looks like all the packages I installed with Homebrew end up in /usr/local/opt directory.

**Reply by mislav** [reply]
I'm not sure what went wrong but it looks like the readline version under `/usr/local/opt/readline` is incompatible with Ruby 2.7.5. You can try installing the latest 2.7.x version to see if it helps things

**Reply by dungtran3864** [reply]
@mislav So I tried to install Ruby 2.7.6 (which is the latest Ruby 2.7.x version), but the build still failed.

Here are the logs: 
[ruby-build.20221107191327.77463.log](https://github.com/rbenv/ruby-build/files/9956476/ruby-build.20221107191327.77463.log)

Also, here are the versions I have:
rbenv: 1.2.0
readline: 8.2.1
gmp: 6.2.1_1

**Reply by mislav** [reply]
@dungtran3864 Is there any difference in the output when you use `brew --prefix readline` and `command brew --prefix readline` and `/opt/homebrew/bin/brew --prefix readline`?

**Reply by mislav** [reply]
To clarify: I have a feeling that your arm64 build of Ruby is trying to link to amd64 (Intel) version of readline.

**Reply by dungtran3864** [reply]
@mislav 
This is what I have from your suggested commands:
```
MacBook-Pro:~ Zung$ brew --prefix readline
/usr/local/opt/readline
MacBook-Pro:~ Zung$ command brew --prefix readline
/usr/local/opt/readline
MacBook-Pro:~ Zung$ /opt/homebrew/bin/brew --prefix readline
-bash: /opt/homebrew/bin/brew: No such file or directory
```

I used to have an Intel MacBook and now I just recently switched to M1 MacBook. I use this special brew command instead of the "normal" brew command to do everything: `arch -x86_64 brew install rbenv ruby-build` or `arch -x86_64 brew --prefix readline`

**Reply by dungtran3864** [reply]
If I use "normal" brew command on M1 MacBook, I get this error

```
MacBook-Pro:~ Zung$ brew install rbenv
Error: Cannot install in Homebrew on ARM processor in Intel default prefix (/usr/local)!
Please create a new installation in /opt/homebrew using one of the
"Alternative Installs" from:
  https://docs.brew.sh/Installation
You can migrate your previously installed formula list with:
  brew bundle dump
```

This answer on GitHub I follow to fix the error above: https://stackoverflow.com/questions/64963370/error-cannot-install-in-homebrew-on-arm-processor-in-intel-default-prefix-usr

**Reply by mislav** [reply]
@dungtran3864 Thanks for confirming. Your "special" brew setup is the cause of the problem. To fix, you can either install Homebrew the proper way for an M1 machine, or you can uninstall readline, libyaml, and gmp from "special" brew.

**Reply by dungtran3864** [reply]
@mislav I managed to fix it now!! Thank you!! 🎉

The problem is that I used to have an old Intel Mac and switching to new M1 Mac means the original Intel Homebrew as well as previously-installed packages don't work correctly.

Here are what I did:
1. Uninstall my current Homebrew completely with this script from official website: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"`. You can check out the script here: /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"
2. Install a fresh Homebrew for M1 Mac with this script: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`. **You must follow any postinstall instructions** at the end of the successful Homebrew installation to configure the shell or $PATH correctly.
3. Reinstall everything required from rbenv, which means reinstalling rbenv, ruby-build, gmp, readline, libyaml, openssl, etc. Here are the list of commands I ran, which you can find on the official website:
`brew install rbenv ruby-build`
`# run this and follow the printed instructions:
rbenv init`
`brew reinstall openssl@1.1 readline libyaml gmp`
`export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"`
4. Close terminal and install Ruby 2.7.6 version instead `rbenv install 2.7.6`. The reason is that I notice Ruby version 2.7.5 still fails but Ruby version 2.7.6 succeeds. Weird.. 😕 I think this is due to Ruby 2.7.6 is the only stable version for 2.7.x, not the other.

**Reply by dungtran3864** [reply]
@mislav Can you troubleshoot why Ruby 2.7.5 version still failed this time (while 2.7.6 is successful), please?

Here are the logs: 
[ruby-build.20221108124934.66099.log](https://github.com/rbenv/ruby-build/files/9963874/ruby-build.20221108124934.66099.log)

Note: I tried to install Ruby 2.7.4 if this build failed only happens to Ruby 2.7.5, but ironically Ruby 2.7.4 also failed to build.

**Reply by dungtran3864** [reply]
Nevermind, I found the reason why Ruby 2.7.5 is failed to install. I forgot to run `rbenv init` and `export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"` before I run `rbenv install 2.7.5`. I'm able to install and use Ruby 2.7.5 now. Thanks! 😊

**Reply by ChimiGit** [reply]
I have done but it isn't working.

**Comment by p19ky** [comment]
I also have this problem and can't figure out why it is failing. I tried the answer from this thread but nothing seems to work.

`
➜  ~ rbenv install 2.6.4
To follow progress, use 'tail -f /var/folders/px/m4dcvbf92jbgwc06pd194nt80000gn/T/ruby-build.20221201131112.46839.log' or pass --verbose
Downloading ruby-2.6.4.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.4.tar.bz2
Installing ruby-2.6.4...

WARNING: ruby-2.6.4 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using readline from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.0.1 using ruby-build 20221124)

Inspect or clean up the working tree at /var/folders/px/m4dcvbf92jbgwc06pd194nt80000gn/T/ruby-build.20221201131112.46839.81u9oR
Results logged to /var/folders/px/m4dcvbf92jbgwc06pd194nt80000gn/T/ruby-build.20221201131112.46839.log

Last 10 log lines:
transform_mjit_header: SKIPPED to transform __darwin_check_fd_set
transform_mjit_header: SKIPPED to transform __sputc
Undefined symbols for architecture x86_64:
  "__mh_execute_header", referenced from:
      _rb_dump_backtrace_with_lines in addr2line.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [libruby.2.6.dylib] Error 1
make: *** Waiting for unfinished jobs....
ln -sf ../../../.ext/include/x86_64-darwin22/rb_mjit_min_header-2.6.4.h include/ruby-2.6.0/x86_64-darwin22/rb_mjit_min_header-2.6.4.h
`

**Comment by TassainRasool** [comment]
user@01HW2267461 ~ % rbenv global 2.7.0
rbenv: version `2.7.0' not installed
user@01HW2267461 ~ % rbenv install 2.7.0
ruby-build: using openssl@1.1 from homebrew
==> Downloading ruby-2.7.0.tar.bz2...
-> curl -q -fL -o ruby-2.7.0.tar.bz2 https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.0.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14.0M  100 14.0M    0     0  2794k      0  0:00:05  0:00:05 --:--:-- 3040k
==> Installing ruby-2.7.0...

WARNING: ruby-2.7.0 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.rbenv/versions/2.7.0" --with-openssl-dir=/opt/homebrew/opt/openssl@1.1 --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+
-> make -j 10

BUILD FAILED (macOS 13.5.1 on arm64 using ruby-build 20240501)

You can inspect the build directory at /var/folders/8b/647dmgl905vbh36v4p_l0h_m0000gq/T/ruby-build.20240510233815.3165.whbK9z
See the full build log at /var/folders/8b/647dmgl905vbh36v4p_l0h_m0000gq/T/ruby-build.20240510233815.3165.log

**Reply by monfresh** [reply]
The oldest version of Ruby 2.7.x that you can install on arm64 (Apple Silicon Macs) is 2.7.2. However, even 2.7.2 can cause issues. In general, it's recommended to use the latest patch version (the 3rd digit) of any given Ruby version. The latest 2.7.x version is 2.7.8. So, instead of trying to install 2.7.0, it's better to install 2.7.8 and then upgrade your project to 2.7.8. Here's a step-by-step guide I wrote that explains [How and Why to Upgrade the Ruby Version in Your Project](https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project).
